fluent/src/langneg.js

```diff
@@ -0,0 +1,246 @@
+
+/**
+ * Likely subtags map from a locale in minimized form to its maximized form.
+ * It's useful for non-intuitive matches like 'en' -> 'en-US' or marking
+ * what's the languge default region and script.
+ */
+const likelySubtags = {
+  'en': 'en-Latn-US',
+  'sr': 'sr-Cyrl-RS',
+  'sr-RU': 'sr-Latn-RU',
+  'az': 'az-Latn-AZ',
+  'az-IR': 'az-Arab-IR',
+  'zh': 'zh-Hans-CN',
+  'zh-GB': 'zh-Hant-GB',
+  'zh-US': 'zh-Hant-US',
+};
+
+/**
+ * Attempts to build a most likely maximum version of the locale id based
+ * on it's minimal version.
+ *
+ * If possible it uses the lookup table above. If this one is missing,
+ * it only attempts to place the language code in the region position.
+ */
+function addLikelySubtags(locale) {
+  if (likelySubtags.hasOwnProperty(locale)) {
+    return likelySubtags[locale];
+  }
+  const loc = parseLocale(locale, true);
+  if (loc[2] === '*') {
+    loc[2] = loc[0].toUpperCase();
+    return loc.join('-');
+  }
+  return null;
+}
+
+/**
+ * Replaces the region position with a range to allow for matches
+ * with the same language/script but from different region.
+ */
+function getRegionRange(locale) {
+  const localeString = parseLocale(locale, true);
+  localeString[2] = '*';
+  return localeString.join('-');
+}
+
+const languageCodeRe = '([a-z]{2,3}|\\*)';
+const scriptCodeRe = '(?:-([a-zA-Z]{2,4}|\\*))';
+const regionCodeRe = '(?:-([A-Z]{2}|\\*))';
+const variantCodeRe = '(?:-([a-z]+|\\*))';
+
+/**
+ * Regular expression splitting locale id into four pieces:
+ *
+ * Example: `en-Latn-US-mac`
+ *
+ * language: en
+ * script:   Latn
+ * region:   US
+ * variant:  mac
+ *
+ * It can also accept a range `*` character on any position.
+ */
+const localeRe = new RegExp(
+`^${languageCodeRe}${scriptCodeRe}?${regionCodeRe}?${variantCodeRe}?$`);
+
+/**
+ * Parses a locale id using the localeRe into an array with four elements.
+ *
+ * If the second argument `range` is set to true, it places range `*` char
+ * in place of any missing piece.
+ *
+ * It also allows skipping the script section of the id, so `en-US` is properly
+ * parsed as `en-*-US-*`.
+ */
+function parseLocale(locale, range = false) {
+  const result = localeRe.exec(locale);
+  if (!result) {
+    return null;
+  }
+
+  const language = result[1];
+
+  let script = range || result[2] === '*' ? '*' : undefined;
+  if (result[2] !== undefined && result[2].length === 4) {
+    script = result[2];
+  }
+
+  let territory = range ? '*' : undefined;
+  if (result[2] !== undefined && result[2].length === 2) {
+    territory = result[2];
+  } else if (result[3] !== undefined) {
+    territory = result[3];
+  }
+
+  let variant = range ? '*' : undefined;
+  if (result[4]) {
+    variant = result[4];
+  }
+
+  return [language, script, territory, variant];
+}
```

Comment:
... this code would become easier, if the regular expression was more standards-conformant, I think.

Also, we should probably normalize casing here, so that we can compare strings later on. 'cause language tags are compared case-independently.